W-18408079-Update-DW-content-with-AI-CP (#457)
mulesoft/docs-code-builder: modules/ROOT/pages/int-configure-dw-expressions.adoc

= Using DataWeave Expressions and Transformations in Anypoint Code Builder
:page-aliases: int-use-dw-to-transform-data.adoc, int-address-dw-errors.adoc, int-import-dw-libraries.adoc

DataWeave is the MuleSoft programming language for data transformation and for defining expressions. Use DataWeave to process Mule event data, such as `payload`, `attributes`, and `vars`, in connector operations and other components. Develop DataWeave expressions and transformations in your Mule applications using the *Expression Field*, *Expression Builder*, *Transformation Builder*, or by editing the configuration XML. Generate DataWeave transformations with AI help by providing input and output sample data or metadata.

image::int-dw-transform-ui.png["A configuration panel showing three numbered callouts: 1) fx button for Expression Field, 2) Expression Builder button, and 3) Transformation Builder button"]

[calloutlist]

> . *Expression Field*: Enter a DataWeave expression or script in a component configuration panel.

Reviewer: ```suggestion
. *Expression Field*: Enter a DataWeave expression or script in a component configuration panel.
```

This entry is weird as it's a command and the others aren't.

Author: Great catch! Thanks!

* Concatenation of simple fields
* `Array` to `Array` (same level)

[[prereq-dw-transforms]]
== Before You Begin

* Understand the functions, operators, variables, and selectors of DataWeave expressions.
+

> xref:dataweave::dw-functions.adoc[DataWeave functions and operators], and data types that their parameters accept. For example, the concatenation function (xref:dataweave::dw-core-functions-plusplus.adoc[]) accepts strings, arrays, objects, and date-related data types. 

Reviewer: This line reads weird. I would just link the reference doc and remove the example.

Author: A link to the source will do! Thanks!

> * Review the built-in documentation for functions the IDE provides. For example, see <<tab-functions>>.

Reviewer: Does this lead into the next bulleted section? It's unclear. The ** bullets real a bit off because of it.

Author: Yes! The structure in the "before you begin" section is made of bullets. The "**" bullets are a subtopic of the "*"  topic. Did I answer your question?

** xref:dataweave::dataweave-variables-context.adoc[Predefined Mule event variables] such as `payload`, `attributes`, `vars`.

** xref:dataweave::dataweave-selectors.adoc[DataWeave selectors] for capturing the values of fields within the payload, an attribute, or a variable within the xref:mule-runtime::about-mule-event.adoc[Mule event]. 
+
For more resources, see xref:dataweave::index.adoc[]. 

> * Make sure you have these permissions to use the Anypoint Code Builder AI feature:

Reviewer: ```suggestion
* Ensure you have these permissions to use the Anypoint Code Builder AI feature:
```

Author: According to Vale, we should use "Make sure" instead of "Ensure". Thanks for your suggestion!

[[preview-results-fx]]
==== Preview Results in the UI

> Preview auto generated sample data, or preview your own sample data. Preview the output in the *Preview* tab of *Expression Builder* for a component expression field.

Reviewer: Your call, but would auto-generated not be correct?

Author: According to our style guide (https://confluence.internal.salesforce.com/spaces/DOCTEAM/pages/143000739/hyphen+Style+Guidelines), we must use hyphen "only if a phrase is ambiguous without a hyphen." I don't think this is the case. Thanks!

On *Transformation Builder*, select the option that best suits your needs:

* <<map-with-ai, Map with AI>>
* <<map-and-transform-with-ai, Map and Transform with AI>>

[[considerations-for-transformation-builder-with-ai]]
=== Considerations for AI-Assisted Transformations

If using AI-assisted transformations, evaluate these considerations:

* Metadata provides the shape of the output.
* Sample data provides the context to generate functions or expressions for calculations, conditions, and other operations. You can provide Web Services Description Language (WSDL) files with sample data. See <<sample-data>> for supported data formats (MIME types) for sample data.
* AI requires valid data to create the transformation script. Write it manually if needed.
* AI chat history isn't available in *Transformation Builder*.
* Avoid using real personally identifiable information (PII) in sample data or AI inputs. Provide input and output sample data that contains only fictional or masked data.

[[map-with-ai]]
=== Map with AI

Get an automatic field-to-field mapping when accessing *Transformation Builder*.

If you don't have access to *Map with AI*, you must:

> * Make sure your configuration meets the prerequisites for using AI-assisted transformations. See <<prereq-dw-transforms>>.

Reviewer: ```suggestion
* Ensure your configuration meets the prerequisites for using AI-assisted transformations. See <<prereq-dw-transforms>>.
```

Author: See my comment above. Thanks!

+
The mapping starts automatically considering the sample data provided. 
+
The mapping is applied to the XML file.
+
The *Preview* tab shows the output of the transformation.
. After the mapping is complete, you must:
.. Click *Apply Mapping* to apply suggested mapping to your project.
.. Click *Revise Mapping* to get new suggestions from the AI.

[[map-and-transform-with-ai]]
=== Map and Transform with AI

Create a DataWeave transformation script assisted by AI. Provide input and output sample data, the input and output metadata, and optional instructions to guide the AI.

. Open your Mule project.
. Select a component that has a *fx* field.
. Open *Transformation Builder*.
+
image::int-map-with-ai.png["Transformation Builder showing 'Fetching metadata' status message with loading indicator for AI-assisted mapping functionality"]
+
The mapping starts automatically considering the data provided. The mapping is applied to the XML file.
. Click *Map and Transform with AI*.
. Provide this information if needed, or modify it as needed:
* *Input sample* data 
* *Output sample* data 
* Optional instructions to guide the AI
[NOTE]

> If modifying the input and output sample data, provide the same information to help the AI understand the intended transformation. The input data can come from payload, attributes, or variables.

Reviewer: ```suggestion
If modifying the input and output sample data, provide the same information to help the AI understand the intended transformation. The input data can come from payloads, attributes, or variables.
```

Not a big deal, but maybe "the payload" or "payloads"

Author: Great catch! Thanks!

. Click *Transform*.
+

> Einstein AI analyzes your input and output samples, and the instructions -if any-, to generate DataWeave script suggestions that aim to perform the suggested transformation. 

Reviewer: ```suggestion
Einstein AI analyzes your input and output samples, and the instructions -if any-, to generate DataWeave script suggestions that aim to perform the suggested transformation. 
```

Is the "-if any-" notation correct?

Author: Yes! Instructions are optional in this process.
